### DISCUSSION TITLE
[workflow component] Using attributes to define a workflow

### DISCUSSION THREAD

**Discussion by tacman** [symfony/symfony#47899] [Q&A] [answered]
Inspired by @kbond 's invokable commands (https://github.com/zenstruck/console-extra), I've been playing around with defining a workflow using attributes.  Since often there's just one workflow for an entity, this example simply annotates the Task entity.

```php
#[Workflow(name: 'task', prefix: 'PLACE_')]
class Task implements MarkingInterface
{
    use MarkingTrait;

    const PLACE_DRAFT = 'draft';
    const PLACE_REVIEWED = 'reviewed';
    const PLACE_REJECTED = 'rejected';
    #[Place(metadata: ['description' => "Published on the website"])]
    const PLACE_PUBLISHED = 'published';

    #[Transition(self::PLACE_DRAFT, self::PLACE_REVIEWED, metadata: ['description' => 'When the draft is ready'])]
    const TRANSITION_REVIEW = 'review';

    #[Transition(from: self::PLACE_REVIEWED, to: self::PLACE_PUBLISHED,
        guard: "is_granted('ROLE_REVIEWER')")]
    const TRANSITION_PUBLISH = 'publish';

    #[Transition([self::PLACE_REVIEWED, self::PLACE_DRAFT], self::PLACE_REJECTED)]
    const TRANSITION_REJECT = 'reject';

    #[Transition([self::PLACE_REVIEWED, self::PLACE_PUBLISHED], self::PLACE_DRAFT)]
    const TRANSITION_RESET = 'reset';
```

The workflow attribute on the class defines the name (defaults to the class shortname), and if a prefix is defined, it gets all the constants that begin with it and uses those to define the places.  Alternatively, the Place attribute can be put on the place names.

The transition attribute defines the from and to, along with guard and metadata.

I'm pleased with how it looks, and wanted to get some feedback from the community before I continue.  

I'm also stuck on a technical issue -- where should these workflows be defined and added to the registry?  In the compiler pass?  This is in a bundle, and even when I process it in the workflow.php configuration file, I'm having issues seeing the bundle trait.

I'll post the specific code once I get some confirmation that I'm on the right path.  Thanks!

**Comment by lyrixx** [comment, marked as answer]
Hello. You can configure the workflow in you bundle extension. See what is done in the symfony bundle extension

About the registry, you should not use it anymore
